src/Timer/Timer.tsx

```diff
@@ -1,83 +1,63 @@
-
-        this.state = {
-            timer: "00:00",
-        };
-    }
-
-    startInterval(timeout: number): void {
-        this.setState({ timer: this.convertMStoTime(timeout) }, () => {
-            this.clearInterval();
-            this.innerInterval = setInterval(() => {
-                if (timeout > 0) {
-                    timeout = timeout - 1000;
-                    this.setState({ timer: this.convertMStoTime(timeout) }, () => {
-                        if (timeout === 0) {
-                            this.props.callback();
-                            this.clearInterval();
-                        }
-                    });
-                }
-            }, 1000);
-        });
-    }
-
-    clearInterval(): void {
-        if (this.innerInterval) {
-            clearTimeout(this.innerInterval);
-            this.innerInterval = null;
-        }
-    }
-
-    convertMStoTime(value: number): string {
-        const date: Date = new Date(value);
-        return (
-            (date.getUTCHours() > 0 ? date.getUTCHours() + ":" : "") +
-            (date.getUTCMinutes() < 10 ? "0" + date.getUTCMinutes() : date.getUTCMinutes()) +
-            ":" +
-            (date.getUTCSeconds() < 10 ? "0" + date.getUTCSeconds() : date.getUTCSeconds())
-        );
-    }
-
-    componentDidMount() {
-        if (this.props.duration !== null && this.props.duration !== undefined) {
-            this.startInterval(this.props.duration);
-        }
-    }
-
-    componentWillUnmount() {
-        this.clearInterval();
-    }
-
-    componentDidUpdate(prevProps: TimerProps): void {
-        if (prevProps.duration !== this.props.duration) {
-            if (this.props.duration !== null && this.props.duration !== undefined) {
-                this.startInterval(this.props.duration);
+import classnames from "classnames";
+import { min } from "lodash";
+
+export type TimerProps = JSX.IntrinsicElements["time"] & {
+    duration?: number;
+    onTimerEnded?: () => void;
+};
+
+export const Timer: React.FC<TimerProps> = React.memo(({ duration, onTimerEnded, ...props }: TimerProps) => {
+    const [time, setTime] = React.useState<string>("00:00");
+    const countDown = React.useRef<number>(0);
+    const timer = React.useRef<any>(0);
+
+    /**
+     * Converts the numeric value into two digits string for display purposes
+     * @param {number} value The numberic value to be converted
+     * @returns The two digits string value
+     */
+    const toDoubleDigits = React.useCallback((value: number): string => {
+        return value < 10 ? `0${value}` : String(value);
+    }, []);
+
+    /** Clears the timer */
+    const clearTimer = React.useCallback((): void => {
+        countDown.current = 0;
+        timer.current && clearInterval(timer.current);
+    }, []);
+
+    /** Kicks off the timer */
+    const startTimer = React.useCallback((): void => {
+        timer.current = setInterval(() => {
+            if (countDown.current--) {
```

Comment:
How does this work ? Find it a little confusing . Isn't -- an operation ?

Reply:
I wanted to leave a comment explaining it, but then I thought it's straightforward.

It will check if the value of `countDown` is more than zero before decrementing it by 1.